src/zcash/circuit/gadget.tcc

```diff
@@ -0,0 +1,349 @@
+    boost::array<std::shared_ptr<digest_variable<FieldT>>, NumInputs> zk_input_macs;
+    boost::array<std::shared_ptr<digest_variable<FieldT>>, NumOutputs> zk_output_commitments;
+    pb_variable_array<FieldT> zk_vpub_old;
+    pb_variable_array<FieldT> zk_vpub_new;
+
+    // Aux inputs
+    pb_variable<FieldT> ZERO;
+    std::shared_ptr<digest_variable<FieldT>> zk_phi;
+    pb_variable_array<FieldT> zk_total_uint64;
+
+    // Input note gadgets
+    boost::array<std::shared_ptr<input_note_gadget<FieldT>>, NumInputs> zk_input_notes;
+    boost::array<std::shared_ptr<PRF_pk_gadget<FieldT>>, NumInputs> zk_mac_authentication;
+
+    // Output note gadgets
+    boost::array<std::shared_ptr<output_note_gadget<FieldT>>, NumOutputs> zk_output_notes;
+
+public:
+    // PRF_pk only has a 1-bit domain separation "nonce"
+    // for different macs.
+    BOOST_STATIC_ASSERT(NumInputs <= 2);
+
+    // PRF_rho only has a 1-bit domain separation "nonce"
+    // for different output `rho`.
+    BOOST_STATIC_ASSERT(NumOutputs <= 2);
+
+    joinsplit_gadget(protoboard<FieldT> &pb) : gadget<FieldT>(pb) {
+        // Verification
+        {
+            // The verification inputs are all bit-strings of various
+            // lengths (256-bit digests and 64-bit integers) and so we
+            // pack them into as few field elements as possible. (The
+            // more verification inputs you have, the more expensive
+            // verification is.)
+            zk_packed_inputs.allocate(pb, verifying_field_element_size());
+            pb.set_input_sizes(verifying_field_element_size());
+
+            alloc_uint256(zk_unpacked_inputs, zk_merkle_root);
+            alloc_uint256(zk_unpacked_inputs, zk_h_sig);
+
+            for (size_t i = 0; i < NumInputs; i++) {
+                alloc_uint256(zk_unpacked_inputs, zk_input_nullifiers[i]);
+                alloc_uint256(zk_unpacked_inputs, zk_input_macs[i]);
+            }
+
+            for (size_t i = 0; i < NumOutputs; i++) {
+                alloc_uint256(zk_unpacked_inputs, zk_output_commitments[i]);
+            }
+
+            alloc_uint64(zk_unpacked_inputs, zk_vpub_old);
+            alloc_uint64(zk_unpacked_inputs, zk_vpub_new);
+
+            assert(zk_unpacked_inputs.size() == verifying_input_bit_size());
+
+            // This gadget will ensure that all of the inputs we provide are
+            // boolean constrained.
+            unpacker.reset(new multipacking_gadget<FieldT>(
+                pb,
+                zk_unpacked_inputs,
+                zk_packed_inputs,
+                FieldT::capacity(),
+                "unpacker"
+            ));
+        }
+
+        // We need a constant "zero" variable in some contexts. In theory
+        // it should never be necessary, but libsnark does not synthesize
+        // optimal circuits.
+        // 
+        // The first variable of our constraint system is constrained
+        // to be one automatically for us, and is known as `ONE`.
```

Comment:
Switch around this comment so that the part about `ONE` doesn't appear to be associated with the `ZERO.allocate(pb);`.